File: 023.pl
#!/usr/bin/perl
use v5.12;
use warnings;

sub sum_of_divisors {
    my $num = shift;
    my ($prime, $sum, $tmp) = (2, 1);
    while ($prime**2 <= $num && $num > 1) {
        if (!($num % $prime)) {
            $tmp = $prime**2;
            $num /= $prime;
            while (!($num % $prime)) {
                $tmp *= $prime;
                $num /= $prime;
            }
            $sum *= $tmp - 1;
            $sum /= $prime - 1;
        }
        if ($prime == 2) { $prime = 3 }
        else { $prime += 2 }
    }
    if ($num > 1) { $sum *= $num + 1 }
    return $sum
}

my $limit = 20162;
my $sum;
my %abn;

foreach my $num (1 .. $limit) {
    if (sum_of_divisors($num) > 2 * $num) { ++$abn{$num} }
    foreach my $x (keys %abn) {
        if (($num - $x) ~~ %abn) {
            $sum += $num;
            last;
        }
    }
}
say (($limit + 1) * $limit / 2 - $sum);

File: 023.py
#!/usr/bin/env python
#-*- coding:utf-8 -*-
from euler import divisor_sum

limit = 20162
ans = 0
abn_nums = set()
for num in xrange(1, limit):
    if divisor_sum(num) > 2 * num:
	abn_nums.add(num)
    if not any( (num - abn in abn_nums) for abn in abn_nums):
	ans += num
 
print(ans)
